### ISSUE TITLE
Audit blocks' implicit 'self' capturing

### ISSUE THREAD

**Issue by joeljfischer** [smartdevicelink/sdl_ios#129]
Accessing iVars within blocks causes an implicit capture of self, which will lead to a retain cycle. The library needs to be audited for these issues. Below is one such:

```objc
-(void) append:(id) toAppend {
    dispatch_async(dispatch_get_main_queue(), ^{
        //Insert the new data
        NSMutableDictionary *dictionary = [NSMutableDictionary dictionary];

        [dictionary setObject:toAppend forKey:@"object"];
        [dictionary setObject:[NSDate date] forKey:@"date"];

        [messageList addObject:dictionary];
        NSIndexPath *newIndex = [NSIndexPath indexPathForRow:(messageList.count - 1) inSection:0];
        [self.tableView insertRowsAtIndexPaths:[NSArray arrayWithObject:newIndex] withRowAnimation:UITableViewRowAnimationBottom];

        //If we were at the bottom, scroll to the new bottom.
        if (atBottom) {
            [self.tableView scrollToRowAtIndexPath:newIndex atScrollPosition:UITableViewScrollPositionBottom animated:YES];
        }

        [self.tableView reloadData];
    });
}
```

(Note this is a bad example, since self doesn't actually capture this block)

Here, both `messageList` and `atBottom` are iVars, which means that their calls actually look like `self->messageList`. This is an implicit capture of self and a retain cycle. One way to resolve this is to create a "weak" self outside of the block using something like `__weak typeof(self) weakSelf = self`, then within the block, recapturing the `weakSelf` as a strong, but scoped variable that will be released at the end of the block. 

Simply using the `weakSelf` could mean that it is released out from under us in the middle of the block. So within the block we would have `__strong typeof(self) strongSelf = weakSelf`. Then to access the iVars, properties, call methods, etc. we would reference the strongSelf, like so: `[strongSelf->messageList addObject:dictionary]`.

**Comment by adein** [comment]
@joeljfischer I think this SO post is related to what you were describing:
[Strong To Weak Reference](https://stackoverflow.com/questions/23806751/strong-reference-to-a-weak-references-inside-blocks)

**Comment by joeljfischer** [comment]
Yup, that describes it. It's also described in WWDC 2012 Session 712. [This links to the PDF](http://adcdownload.apple.com/wwdc_2012/wwdc_2012_session_pdfs/session_712__asynchronous_design_patterns_with_blocks_gcd_and_xpc.pdf) which you can only see logged in to Apple's dev site. See page 105.

**Comment by joeljfischer** [comment]
Fixed in #316
